proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+to the user if certain conditions decided by dapp developers are met.
+
+### Other terminology
+
+`Owner` of an account: It is the account owner as specified in the `owner`
+field in the account structure. In the case of an externally owned account
+(i.e, keypair) owner is a `system program` and in case of a Program derived
+address owner is usually the program.
+
+Account `Authority`: The true authority for the account. Let's take an example
+of associated token account. A associated token account is a `PDA` derived
+from associated token program, where the owner is token program. But the
+program internally saves who has the authority over the account. Only token
+program can change the data of a token account. For operations like
+withdrawing tokens, the authority of the associated token account has to be
+the signer for the transaction containing transfer instruction. To receive
+tokens from another account, the authority's signature is not required.
+
+`Other Fees`: Fees other than application fees
+`base fees + prioritization fees`.
+
+## Detailed Design
+
+Application fees enable dapp developer to decide who to rebate and who to
+penalize through runtime features. In particular, programs needs to be able
+to penalize, even if the application is not CPI'd to.
+
+The checks on the application fees will be taken care of by Solana runtime.
+Total application fees paid will be included in the transaction through a
+special `PayApplicationFees` similar to the `ComputeBudget` program so that it
+is easier to calculate the total amount of fees that will be paid. In case of
+nonpayment or partial payment, the program is never executed. This instruction
+cannot be CPI'ed into.
+
+Application developers will be able to rebate these fees through a new runtime
+mechanism, but only if the transaction actually succeeds, this way CU heavy
+simulation of the program while write-locking the respective accounts will
+require full fee payment.
+
+### Payment
+
+When the cluster receives a new transaction, `PayApplicationFees` instruction is
+decoded to calculate the total fees required for the transaction message. Then
+we verify that the fee-payer has a minimum balance of:
+`per-transaction base fees` + `prioritization fees` +
+`maximum application fees to be paid`
+
+If the payer does not have enough balance, the transaction is not scheduled
+and fails with an error `InsufficientFunds`. Consider this case the same as if
+the payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+application fees. For all accounts with application fees, the fees paid should
+be greater than the required fees. In case of overpayment, the difference is
+stored in a variable and will be paid back to the payer in any case. In case the
+application fees are insufficiently paid or not paid then the transaction fails
+with an error `ApplicationFeesNotPaid`. The payer will pay just `other fees`.
```

Comment:
If returns `TransactionError::ApplicationFeesNotPaid` during payer account validation, the payer still need to pay `other fees` because accounts were loaded?

Reply:
Yes, it is similar case to if the transaction failed for some reason. Just because validator loaded the accounts payer has to pay the base fees.

Reply:
thanks for clarifying